Add a containerd module

vdemeester opened the pull request:
This add a containerd module so you can have a containerd systemd service
running and using it with other tools. One possibility is to make dockerd
using this containerd service using `--containerd` flag on dockerd

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

###### Motivation for this change

Have a `containerd` module, that could be used with `dockerd` or `kubernetes` and `cri-containerd`, etc..

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Mic92 commented:
the module is not included in `modules/module-list.nix`

vdemeester commented:
@Mic92 oups.. I forgot to commit that 🤦‍♂️

vdemeester commented:
ping 👼 🙏

Mic92 commented on nixos/modules/virtualisation/containerd.nix:
```diff
@@ -0,0 +1,84 @@
+            ""
+            ''
+              ${cfg.package}/bin/containerd \
+                ${cfg.extraOptions}
```
somebody found that trick!

vdemeester replied:
:thinking: :stuck_out_tongue_closed_eyes:

Mic92 commented on nixos/modules/virtualisation/containerd.nix:
```diff
@@ -0,0 +1,84 @@
+
+      systemd.services.containerd = {
+        wantedBy = [ "multi-user.target" ];
+        environment = proxy_env;
```
Where is this option defined?

vdemeester replied:
Oups, it's not supposed to be there ... :sweat:

vdemeester commented:
Updated :angel:

Mic92 commented on nixos/modules/virtualisation/containerd.nix:
```diff
@@ -0,0 +1,83 @@
+
+  config = mkIf cfg.enable (mkMerge [{
+      environment.systemPackages = [ cfg.containerd ];
+      systemd.packages = [ cfg.containerd ];
```
This should be `pkgs.containerd` or `cfg.package`. Have you ever tested this module?

Mic92 commented on nixos/modules/virtualisation/containerd.nix:
```diff
@@ -0,0 +1,83 @@
+          ExecStart = [
+            ""
+            ''
+              ${cfg.package}/bin/containerd \
```
`cfg.package` does not exists as well.

Mic92 commented:
Please test modules before ticking the `tested` checkbox in the pull request templates.

vdemeester commented:
@Mic92 oh.. seems like I didn't correctly update that file frol my nixos configuration, sorry 🙇, I'll update 😇

colemickens commented:
This seems okay to merge now?

vdemeester commented:
@colemickens I need to updated the module to be able to add some additionnal `packages` in the service path… This will be useful when using `containerd` with additionnal runtime (other than `runc`)

vdemeester commented:
Updated :angel:

colemickens commented:
Ping @Mic92, can you review/merge? Thanks!

vdemeester commented:
ping @Mic92 @nlewo

PsyanticY commented:
ping @Mic92

JohnAZoidberg commented on nixos/modules/virtualisation/containerd.nix:
```diff
@@ -0,0 +1,91 @@
+      default = ["/run/containerd/containerd.sock"];
+      description =
+      ''
+        A list of unix and tcp containerd should listen to. The format follows
```
A list of unix and tcp what? Sockets?

```suggestion
        A list of unix and TCP sockets containerd should listen to. The format follows
```

vdemeester commented on nixos/modules/virtualisation/containerd.nix:
```diff
@@ -0,0 +1,91 @@
+          ${cfg.extraOptions}
+        ''];
+      };
+      path = [ pkgs.containerd ] ++ cfg.packages;
```
I think it is required to get `containerd-shim` and some other `containerd-*` commands (mainly the shims)

vdemeester committed:
Add a containerd module

This add a containerd module so you can have a containerd systemd service
running and using it with other tools. One possibility is to make dockerd
using this containerd service using `--containerd` flag on dockerd

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented:
@Mic92 @nlewo @JohnAZoidberg finally updated this PR :sweat_smile:

colemickens commented:
@vdemeester did you abandon containerd or do you have this out-of-nixpkgs somewhere now?

vdemeester commented:
> @vdemeester did you abandon containerd or do you have this out-of-nixpkgs somewhere now?

I keep a module in my own configuration, but I use it less than I used to. I closed the PR because I don't find enough time right now to update it and get it merge :sweat_smile:. If someone wants to take over that would be nice, otherwise, the day I find some time, I'll re-open a new one :wink:

vdemeester commented:
I can also re-open if we think it can get merged as is :wink:
